### ISSUE TITLE
categorical twists with arrow (invalidation of valid DictionaryArray, wrong minimization of ChunkedArray, option ChunkedArray are non-categorical)

### ISSUE THREAD

**Issue by drahnreb** [scikit-hep/awkward#674]
Hi Jim, there are still problems with categoricals. This is a new issue but it is related to #400 and PR #403, which we discussed a few months ago.

When reading in partitioned parquets (e.g. by `year`, which produce categoricals, similar to 3.) I encountered several problems with categoricals debugging. Data doesn't seem to be corrupted but e.g. writing to parquet is affected.

I combined three bugs in one issue as they might be partly linked to each other and be addressed in a joint way. If you disagree please feel free to split the following:
1. Invalidation of valid categorical arrays
2. Minimization fails for `ChunkedArray` with categoricals using `ak.from_arrow()`
3. Option-typed `DictionaryArray` are treated as non-categorical awkward arrays

#### Minimal reproducible examples:
```python
>>> import awkward as ak
>>> import pyarrow as pa
>>> (ak.__version__, pa.__version__) # pyarrow 2.0 should fix ARROW-9801 and DictionaryArray indices look good
('1.0.2', '2.0.0')
```

**1.**
_Invalidation of valid categorical arrays_
```python
>>> original = ak.Array([2019, 2020, 2021, 2020, 2019])
>>> ak.is_categorical(original)
>>> categorical = ak.to_categorical(original)
>>> categorical.layout # looks fine
<IndexedArray64>
     <parameters>
         <param key="__array__">"categorical"</param>
     </parameters>
     <index><Index64 i="[0 1 2 1 0]" offset="0" length="5" at="0x56393ae35f60"/></index>
     <content><NumpyArray format="l" shape="3" data="2019 2020 2021" at="0x56393ab84550"/></content>
</IndexedArray64>
>>> ak.is_categorical(categorical) # is fine
True
>>>  ak.categories(categorical) # does fine
<Array [2019, 2020, 2021] type='3 * int64'>
>>> ak.is_valid(categorical) # but is invalid
False
>>> ak.is_valid(original)
True
```

**2.**
_Minimization fails for `ChunkedArray` with categoricals using ak.from_arrow()_
```python
>>> indices = pa.array([0, 1, 0, 1, 2, 0, 2])
>>> dictionary = pa.array([2019, 2020, 2021])
>>> dict_array = pa.DictionaryArray.from_arrays(indices, dictionary)
>>> # create a table with 3 chunkedarrays
>>> batch = pa.RecordBatch.from_arrays([dict_array], ['year'])
>>> batches = [batch] * 3
>>> table = pa.Table.from_batches(batches)
>>> table['year']
<pyarrow.lib.ChunkedArray object at 0x7fbb02b05310>
[

  -- dictionary:
    [
      2019,
      2020,
      2021
    ]
  -- indices:
    [
      0,
      1,
      0,
      1,
      2,
      0,
      2
    ],

  -- dictionary:
    [
      2019,
      2020,
      2021
    ]
  -- indices:
    [
      0,
      1,
      0,
      1,
      2,
      0,
      2
    ],

  -- dictionary:
    [
      2019,
      2020,
      2021
    ]
  -- indices:
    [
      0,
      1,
      0,
      1,
      2,
      0,
      2
    ]
]
>>> array = ak.from_arrow(table)
>>> ak.is_valid(array), ak.validity_error(array)
(False,
 'at layout.field(0).content (IndexedArray64): __array__ = "categorical" requires contents to be unique')
>>> array.layout # this really doesn't seem to look fine
<RecordArray>
     <field index="0" key="year">
         <IndexedOptionArray64>
             <index><Index64 i="[0 1 2 3 4 ... 16 17 18 19 20]" offset="0" length="21" at="0x55a1f6baaa10"/></index>
             <content><IndexedArray64>
                 <parameters>
                     <param key="__array__">"categorical"</param>
                 </parameters>
                 <index><Index64 i="[0 1 0 1 2 ... 6 7 8 6 8]" offset="0" length="21" at="0x55a1f6b06cc0"/></index>
                 <content><NumpyArray format="l" shape="9" data="2019 2020 2021 2019 2020 2021 2019 2020 2021" at="0x55a1f69961d0"/></content>
             </IndexedArray64></content>
         </IndexedOptionArray64>
     </field>
</RecordArray>
```

**3.**
_Option-typed `DictionaryArray` are treated as non-categorical awkward arrays_
Not sure if intended, but I couldn't figure why this behavior would be necessary:
```python
>>> nan_indices = pa.array([0, 1, 0, 1, 2, None, 0, 2])
>>> option_dict_array = pa.DictionaryArray.from_arrays(nan_indices, dictionary)
>>> option_awk_array = ak.from_arrow(option_dict_array)
>>> option_awk_array.layout
<IndexedOptionArray64>
    <index><Index64 i="[0 1 0 1 2 -1 0 2]" offset="0" length="8" at="0x55a1f68d52f0"/></index>
    <content><NumpyArray format="l" shape="3" data="2019 2020 2021" at="0x7fbb09a0a6c0"/></content>
</IndexedOptionArray64>
```
However, this might be the only case where one could argue that `ak.is_valid()` would rightfully claim non-uniqueness (is `None` unique?). But it checks valid, as it is missing the param keys:
```python
<parameters>
    <param key="__array__">"categorical"</param>
</parameters>
```

**Comment by jpivarski** [comment]
I thought we didn't support partitioned Parquet datasets yet (#368). For issue 1, you can get more information using [ak.validity_error](https://awkward-array.readthedocs.io/en/latest/_auto/ak.validity_error.html) to get the error message. Issue 2 is likely something that I just didn't know about: is that a ChunkedArray with 3 chunks? The Arrow → Awkward function doesn't attempt to consolidate dictionaries from different chunks, so they could get duplicated, and we'll have to add code to stop that. Issue 3 is probably an oversight due to the fact that IndexedOptionArray does both indexing and option-type, but the test for categoricalness might only be checking for IndexedArrays.

These sound like not-too-difficult problems, all of them to be addressed on the Python level, but most will require me to become more knowledgeable about Arrow. You're right that they should probably be dealt with all at once, and probably along with #368 at the same time.

**Comment by drahnreb** [comment]
> I thought we didn't support partitioned Parquet datasets yet (#368).

Yes correct, thanks for pointing that out. I missed that. But it doesn't seem to be the root cause for Issue 1 and 3.



> For issue 1, you can get more information using [ak.validity_error](https://awkward-array.readthedocs.io/en/latest/_auto/ak.validity_error.html) to get the error message.

Yes, I looked into that, result is `'at layout (IndexedArray32): __array__ = "categorical" requires contents to be unique'` which doesn't seem to be true here as `categorical` is correctly minimized. I've not yet looked further into the code of `ak.is_valid()`...



> Issue 2 is likely something that I just didn't know about: is that a ChunkedArray with 3 chunks? ... we'll have to add code to stop that.

Yes, a) it is a collection of primitives ... b) and could be [flattened](https://arrow.apache.org/docs/python/generated/pyarrow.ChunkedArray.html#pyarrow.ChunkedArray.combine_chunks). The structure is used when dealing with partitioned parquets in particular.


> These sound like not-too-difficult problems, all of them to be addressed on the Python level, but most will require me to become more knowledgeable about Arrow.

I try looking into it by end of the week.

**Comment by drahnreb** [comment]
Resolved via #727
